io/src/main/java/io/smallrye/common/io/FileUtils.java

```diff
@@ -0,0 +1,565 @@
+     *
+     * @param file to be checked, must be not null
+     * @return the passed file if it is a directory
+     * @throws IllegalArgumentException if the path is {@code null}
+     * @throws UncheckedIOException if is not a directory or determination of
+     *         type fails
+     */
+    public static File checkIsDirectoryFollowLinks(final File file) throws UncheckedIOException {
+        checkNotNullParam("file", file);
+        return PathUtils.doCheckIsDirectoryFollowLinks(file.toPath()).toFile();
+    }
+
+    /**
+     * A fluent style directory check, not following links, returning the passed
+     * argument if path is a directory. Check is done exactly as
+     * {@link java.nio.file.Files#isDirectory(Path, LinkOption...)} but with
+     * explicit null check and unchecked {@code java.io.UncheckedIOException}.
+     *
+     * @param file to be checked, must be not null
+     * @return the passed file if it is a directory
+     * @throws IllegalArgumentException if the path is {@code null}
+     * @throws UncheckedIOException if is not a directory or determination of
+     *         type fails
+     */
+    public static File checkIsDirectoryNoFollowLinks(final File file) throws UncheckedIOException {
+        checkNotNullParam("file", file);
+        return PathUtils.doCheckIsDirectoryNoFollowLinks(file.toPath()).toFile();
+    }
+
+    /**
+     * A fluent style regular file check, following links, returning the passed
+     * argument if path is a directory. Check is done exactly as
+     * {@link java.nio.file.Files#isDirectory(Path, LinkOption...)} but with
+     * explicit null check and unchecked {@code java.io.UncheckedIOException}.
+     *
+     * @param file to be checked, must be not null
+     * @return the passed file if it is a regular file
+     * @throws IllegalArgumentException if the path is {@code null}
+     * @throws UncheckedIOException if is not a regular file or determination of
+     *         type fails
+     */
+    public static File checkIsRegularFileFollowLinks(final File file) throws UncheckedIOException {
+        checkNotNullParam("file", file);
+        return PathUtils.doCheckIsRegularFileFollowLinks(file.toPath()).toFile();
+    }
+
+    /**
+     * A fluent style regular file check, not following links, returning the passed
+     * argument if path is a directory. Check is done exactly as
+     * {@link java.nio.file.Files#isDirectory(Path, LinkOption...)} but with
+     * explicit null check and unchecked {@code java.io.UncheckedIOException}.
+     *
+     * @param file to be checked, must be not null
+     * @return the passed path if it is a regular file
+     * @throws IllegalArgumentException if the path is {@code null}
+     * @throws UncheckedIOException if is not a regular file or determination of
+     *         type fails
+     */
+    public static File checkIsRegularFileNoFollowLinks(final File file) throws UncheckedIOException {
```

Comment:
I noticed that some methods take `*Options*` flags and some do not (i.e. `CopyOptions` vs `LinkOptions`); is there a reason for the inconsistency?

Reply:
Yes, this is intended. (No)FollowLinks is important to understand erroneous behaviour, this makes it directly visible in stackstraces and the developer will care, because even autocompletion shows directly the alternatives.
The copy options with it's multiple combinations would blow the code, I assume someone who is looking for "replace" instead of getting a "FileAlreadyExistsException" will start to look at the docs or google stackoverflow.
I have documented the approach in the class javadoc, including links.